src/libtest/lib.rs

```diff
@@ -1084,12 +1084,16 @@ impl MetricMap {
 /// elimination.
 ///
 /// This function is a no-op, and does not even read from `dummy`.
+#[cfg(not(all(target_os = "nacl", target_arch = "le32")))]
 pub fn black_box<T>(dummy: T) -> T {
     // we need to "use" the argument in some way LLVM can't
     // introspect.
     unsafe {asm!("" : : "r"(&dummy))}
     dummy
 }
+#[cfg(all(target_os = "nacl", target_arch = "le32"))]
+#[inline(never)]
+pub fn black_box<T>(dummy: T) -> T { dummy }
```

Comment:
I think this should be an LLVM intrinsic; by taking a reference to `dummy`, `rustc` inserts an alloca+store+load for it and passes the alloca to the asm call. While the alloca+store+load would normally be removed by mem2reg, because of the inline asm (which requires access to a pointer to the value), it can't because the asm reference forces the pointer value to be live. Also, just reading `dummy` (ie not `&dummy`) doesn't work either because constprop (in my earlier example) will do its thang (with sass), much to our chagrin.

I'll be sending a email to llvm-dev proposing `llvm.black_box` Soon(tm).